cli/src/commands/commit.rs

```diff
@@ -54,6 +54,8 @@ pub(crate) fn cmd_commit(
         .get_wc_commit_id()
         .ok_or_else(|| user_error("This command requires a working copy"))?;
     let commit = workspace_command.repo().store().get_commit(commit_id)?;
+    let advanceable_branches = workspace_command
+        .get_advanceable_branches(&**workspace_command.repo(), commit.parent_ids());
```

Comment:
Git user might prioritize the first parent, but I'm not sure. It's unclear whether the branches of a merge commit should advance.

Reply:
The Git CLI represents a merge as a "directed" operation, merging one branch "into" another. The branch beingmmerged "into" is advanced; the other is not. The closest analogue here, I think, would be to advance any branches in `@-` but not in other merge targets. That introduces a dependency on the current working commit state, though, which I think would be weird when making a merge with `new`. Arguably there should be a convenience command that's closer to `git merge` anyway, though; in particular, there's currently no easy way to express "create a new commit only if there are commits I don't already have on my branch", which is the default `git` behavior (and very convenient, especially since it makes `git pull origin main` idempotent).

Reply:
> advance any branches in `@-` but not in other merge targets.

My concern here is what to do if the current working-copy commit is a merge. `@-` points to multiple parents. Perhaps, git will move the first parent branch (because `HEAD` can't point to multiple revisions)?

Reply:
I thought of an alternative. What if we don't advance any branches if there are multiple candidates and instead print a warning for the user indicating that no branches were advanced because multiple candidates were found.

Reply:
Thinking about this again: the typical workflow and desired behavior is something like:

```bash
# "merge from" the branch you want (which will often be `main`); starting from a clean working copy:
> jj new @- <other-branch>
# Now `@` is a merge commit; in Git, the branch that previously pointed to `@-` would now point to `@`
> jj commit -m "Merge from <other-branch> into <current-branch>"
```

If `jj new` with multiple branches doesn't advance at least one of those branches, then you end up with ambiguity. But when `jj new @- <other>` is executed, it's obvious which branch should be advanced (the one pointing to `@-`).

So I wonder if `new @- <others>` should be treated as a special case by advancing the branch from `@-`.

(Again, though, I think it also highlights the usefulness of a separate `merge` command that would do both of the above steps, creating the merge commit _and_ moving the working copy to a new, non-merge commit.)

Reply:
I like this suggestion, but it's worth noting that it introduces a slight discrepancy into the model. Currently `jj new` and `jj commit` create a new commit on top of the target and advance the branch pointing to the target's parent to the target, not to the new commit.

Your suggestion is to make merges a special case and advance the branch from the first target to the new commit if a merge commit is created. That seems reasonable to me and better than doing nothing and making the user resolve it. The reason I think this is ok is because a merge commit is usually "finished" as soon as it's created and can be pushed to a remote. A new empty commit isn't usually "finished" and can't or won't be pushed to a remote.

It would basically work like this:
```
jj log
◉  branch2
│ @  branch1
├─╯
◉
jj new -m "merge commit" branch1 branch2
jj log
@    branch1 merge commit
├─╮
│ ◉
◉ │  branch2
├─╯
◉
```

If Yuya agrees I think we should go with this design and I can update the PR.

Reply:
> Your suggestion is to make merges a special case and advance the branch from the first target to the new commit if a merge commit is created. That seems reasonable to me and better than doing nothing and making the user resolve it. The reason I think this is ok is because a merge commit is usually "finished" as soon as it's created and can be pushed to a remote. A new empty commit isn't usually "finished" and can't or won't be pushed to a remote.
> 
> It would basically work like this:
> 
> ```
> jj log
> ◉  branch2
> │ @  branch1
> ├─╯
> ◉
> jj new -m "merge commit" branch1 branch2
> jj log
> @    branch1 merge commit
> ├─╮
> │ ◉
> ◉ │  branch2
> ├─╯
> ◉
> ```

Suppose the user did `jj new` to finish the merge, doesn't it move the branch2 to the merge commit?
```
jj new @
jj log
@
◉    branch1 branch2 merge commit
├─╮
│ ◉
◉ │
├─╯
◉
```

That was my original concern, and I would argue against adding another special case. If we need to support Git-like merge behavior, I think it's better to add a separate merge command something like `jj git-like-merge <arg>` == `jj new @ <arg>; jj new; jj branch set -r@- ..`.

Reply:
This feature probably shouldn't _ever_ advance a branch from `@-` to `@` if `@` already has a branch pointing to it. I don't think that's really a special case. 

Side-note: IMO there's nothing especially "git-like" about distinguishing between "merge from" and "merge into".

Reply:
> This feature probably shouldn't _ever_ advance a branch from `@-` to `@` if `@` already has a branch pointing to it. I don't think that's really a special case.

I meant this part is the special case: "advance the branches from the first target *to the new commit* if a merge commit is created." If the new commit isn't a merge, `jj new` will advance the branch *to the parent* of the new commit.

> Side-note: IMO there's nothing especially "git-like" about distinguishing between "merge from" and "merge into".

Git has a notion of the current branch (or symbolic `HEAD`.) I think that's the main difference. `jj new` creates a merge of arbitrarily-chosen commits, whereas `git merge` will merge the current commit with the others.

Reply:
> This feature probably shouldn't ever advance a branch from @- to @ if @ already has a branch pointing to it. I don't think that's really a special case.

Currently it does, but maybe you're right and it shouldn't.

> I meant this part is the special case: "advance the branches from the first target to the new commit if a merge commit is created." If the new commit isn't a merge, jj new will advance the branch to the parent of the new commit.

Yep, it's a special case. I don't think that's a big deal though. It should be obvious that a merge commit and a new empty commit are pretty different. I don't think it will be surprising to users that the feature behaves this way after they've seen it once, and I think it makes the feature more useful, even if it's slightly more complex as a result.

Reply:
@yuja When you said "I would argue against another special case", I thought you meant that there would be two special cases involved in handling merges: one when advancing the branch during the merge itself, and one when advancing past the merge. I just meant that with the additional rule of "don't auto-advance to something that's already a different branch", there's only one special case here, not two; but I do agree that the idea we're discussing is a special case.

@emesterhazy Yes, I should have clarified that I am suggesting "don't auto-advance to an existing branch" as a new change.

Reply:
> It should be obvious that a merge commit and a new empty commit are pretty different.

I originally thought that way when I switched from `git`/`hg` to `jj`, but I changed mind. So I think the proposed behavior would be surprising to people who learned `jj` a lot.

Reply:
I have to say, I still don't understand the "jj way", then. Git is effectively the only usable backend outside of Google, and to use git, you need one of:

1. A fork for everything, so you can commit directly to `main`
2. A habit of using `push -a` and later cleaning up after yourself
3. Branches

With branches, there's absolutely a difference between merge commits and non-merge commits. So how *should* I be thinking about this?

Reply:
My current opinion is that this feature already operates on the margins of what might be considered the "native" `jj` workflow. Given that, I think it's ok if it's a little impure or surprising to `jj` experts as long as it is reasonable, consistent, and useful.

This feature is experimental, so we could submit it as-is and request feedback from potential users in #2338. I think we want to resolve these questions:

1. Should we leave the branches as-is if there are multiple branches eligible to advance?
2. Should we only advance a branch if there isn't already a branch pointing to the target commit?
3. Should we add a special case for new commits with multiple parents (aka merges) at moves a branch to the merge commit?

I think my vote for all three of these is yes. Maybe only number 3 is controversial though? If there's agreement on 1 and 2 then I'd like to implement those before I finish https://github.com/martinvonz/jj/pull/3405 and add information about this feature to the changelog.

Reply:
> 1. Should we leave the branches as-is if there are multiple branches eligible to advance?

(clarification: multiple parents, not multiple named branch refs in a single parent?)

> 2. Should we only advance a branch if there isn't already a branch pointing to the target commit?
> 3. Should we add a special case for new commits with multiple parents (aka merges) at moves a branch to the merge commit?

My vote: 1. Maybe?, 2. Maybe?, 3. No.

I hope there will be as few special cases as possible, and to me, 3 is the weirdest one.

However, I'm not the target audience of this feature. I don't use named branches locally, and named branches are created only by `jj git push -c`.

Reply:
@emesterhazy Could you please implement #2, at least? Currently, as far as I can tell, there's no way to pick which branch to advance, so I have to either explicitly push one of them backwards after `jj new`, or, I suppose, temporarily muck with my config.

Another idea is to not advance anything when the user runs `jj new <branchname>`. This would be a more explicit version of not advancing branches when there's already a branch at the destination.

Reply:
I think we should merge this PR now. I've been sitting on it for a while since I'm not sure if this is the best long-term solution for branches. There are [other discussions](https://github.com/martinvonz/jj/issues/3505) and proposals for how branching/bookmarking/topics/aliases etc should work. But, since this is meant to be experimental, I think we can submit it and rip it out later if we decide on a different direction.

---

RE: Implementing number 2, doing this the way I wanted ended up being a little tricky for `commit` since any existing branches are moved to the rewritten `@` commit after the branches are advanced for `advance-branches`.

I have a WIP PR, but I haven't completely gotten it working yet: https://github.com/martinvonz/jj/pull/3548